src/pymor/discretizations/iosys.py

```diff
@@ -0,0 +1,1702 @@
+    """Base class for input-output systems."""
+
+    def __init__(self, m, p, cont_time=True, cache_region='memory', name=None, **kwargs):
+        self.m = m
+        self.p = p
+        super().__init__(cache_region=cache_region, name=name, **kwargs)
+        self.cont_time = cont_time
+
+    def _solve(self, mu=None):
+        raise NotImplementedError
+
+    def __add__(self, other):
+        """Add two input-state-output systems."""
+        assert type(self) == type(other)
+        assert self.cont_time == other.cont_time
+
+        def add_operator(op, other_op):
+            if op.source.id == 'INPUT':
+                if op.range.id == 'OUTPUT':
+                    return BlockOperator([[(op + other_op).assemble()]], source_id=op.source.id, range_id=op.range.id)
+                elif op.range.id == 'STATE':
+                    return BlockOperator.vstack((op, other_op), source_id=op.source.id, range_id=op.range.id)
+                else:
+                    raise NotImplementedError
+            else:
+                if op.range.id == 'OUTPUT':
+                    return BlockOperator.hstack((op, other_op), source_id=op.source.id, range_id=op.range.id)
+                elif op.range.id == 'STATE':
+                    return BlockDiagonalOperator((op, other_op), source_id=op.source.id, range_id=op.range.id)
+                else:
+                    raise NotImplementedError
+
+        new_operators = {k: add_operator(self.operators[k], other.operators[k]) for k in self.operators}
+
+        return self.with_(operators=new_operators, cont_time=self.cont_time)
+
+    def __neg__(self):
+        """Negate input-state-output system."""
+        new_operators = {k: (op * (-1)).assemble() if op.range.id == 'OUTPUT' else op
+                         for k, op in self.operators.items()}
+
+        return self.with_(operators=new_operators)
+
+    def __sub__(self, other):
+        """Subtract two input-state-output system."""
+        return self + (-other)
+
+    def eval_tf(self, s, mu=None):
+        """Evaluate the transfer function."""
+        raise NotImplementedError
+
+    def eval_dtf(self, s, mu=None):
+        """Evaluate the derivative of the transfer function."""
+        raise NotImplementedError
+
+    @cached
+    def bode(self, w):
+        """Evaluate the transfer function on the imaginary axis.
+
+        Parameters
+        ----------
+        w
+            Frequencies at which to compute the transfer function.
+
+        Returns
+        -------
+        tfw
+            Transfer function values at frequencies in `w`,
+            |NumPy array| of shape `(self.p, self.m, len(w))`.
+        """
+        if not self.cont_time:
+            raise NotImplementedError
+
+        return np.dstack([self.eval_tf(1j * wi) for wi in w])
+
+    @classmethod
+    def mag_plot(cls, sys_list, w, plot_style_list=None, ord=None, dB=False, Hz=False):
+        """Draw the magnitude Bode plot.
```

Comment:
Wouldn't be a free function more appropriate? `cls` does not seem to be used at all.

Reply:
Good observation.
I'm actually not so happy with the current implementation. It is too much in the MATLAB style. From what I understand, the [prefered style in matplotlib](https://matplotlib.org/tutorials/introductory/usage.html#coding-styles) is for plotting functions to accept `ax` as an argument and not to create a figure object by itself. Also, maybe it's better to have `mag_plot` as a member function, so instead of
```python
LTISystem.mag_plot([sys1, sys2], w, ['-', '--'], ord=2, dB=True, Hz=True)
```
do
```python
fig, ax = plt.subplots()
kwargs = {'ord': 2, 'dB': True, 'Hz': True}
sys1.mag_plot(ax, w, '-', **kwargs)
sys2.mag_plot(ax, w, '--', **kwargs)
```
@sdrave Any thoughts on this?

Reply:
Yeah, this is looking nicer. In addition, I would make the 'ax' argument optional, and use `matplotlib.pyplot.gca` in case it's not provided.
Do you want to make the changes right now or shall we open a separate issue?

Reply:
Hmm, I see two options: a) Live provide the arguments with every call and live with the fact that you get nonsene if you use different arguments. b) Keep the current style but make it a free function. In addition, there could be a `mag_plot` method which just calls the free function with `[self]` as first argument.

Reply:
`mag_plot` is now a member function. I left `Hz` and `dB` as they were.